File: LibraryImageLoader/src/main/java/com/android/library/CoreImageLoaderUtil.java
package com.android.library;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.widget.ImageView;

import androidx.annotation.Nullable;
import androidx.core.graphics.drawable.RoundedBitmapDrawable;
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory;

import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.BitmapImageViewTarget;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.target.Target;
import com.bumptech.glide.request.transition.Transition;

import java.io.File;


/**
 * 图片加载工具类
 *
 * @author lixurong
 * @date 2017/9/7
 */

public class CoreImageLoaderUtil {
    /* -------------------------接口---------------------------- */

    public interface LoadingImageListener {
        /**
         * 图片加载成功
         *
         * @param t
         * @param <T>
         */
        <T> void onResourceReady(T t);

        /**
         * 加载失败
         */
        void onLoadFailed();
    }

    /**
     * ----------------------------加载图片 int---------------------------
     */
    public static void loadingImg(Context mContext, int resouse, ImageView img,
                                  int placeholder) {
        RequestOptions options = new RequestOptions().placeholder(placeholder)
                .error(placeholder)
                .dontAnimate()
                .diskCacheStrategy(DiskCacheStrategy.NONE);
        GlideApp.with(mContext).load(resouse).apply(options).into(img);
    }

    /**
     * ----------------------------加载本地gif图片---------------------------
     */
    public static void loadingGifImg(Context mContext, int rid, ImageView img) {
        GlideApp.with(mContext).clear(img);
        RequestOptions options = new RequestOptions()
                .dontAnimate()
                .diskCacheStrategy(DiskCacheStrategy.NONE);
        GlideApp.with(mContext).asGif().load(rid).apply(options).into(img);
    }
    /**
     * ----------------------------加载本地sd卡图片---------------------------
     */
    public static void loadingLocalImg(Context mContext, String path, final ImageView img, final int placeholder, final LoadingImageListener listener) {
        RequestOptions options = new RequestOptions().placeholder(placeholder)
                .error(placeholder)
                .dontAnimate()
                .diskCacheStrategy(DiskCacheStrategy.NONE);
        GlideApp.with(mContext).asBitmap().load(path).apply(options).into(new SimpleTarget<Bitmap>() {
            @Override
            public void onResourceReady(Bitmap resource, Transition<? super Bitmap> transition) {
                if(listener != null){
                    listener.onResourceReady(resource);
                }else{
                    img.setImageBitmap(resource);
                }
            }

            @Override
            public void onLoadFailed(@Nullable Drawable errorDrawable) {
                super.onLoadFailed(errorDrawable);
                if(listener != null){
                    listener.onLoadFailed();
                }else{
                    img.setImageResource(placeholder);
                }
            }
        });
    }

    /**
     * ----------------------------加载图片 Bitmap---------------------------
     */
    public static void loadingImg(Context mContext, Bitmap bitmap,
                                  ImageView img, int placeholder) {
        GlideApp.with(mContext).clear(img);
        RequestOptions options = new RequestOptions().placeholder(placeholder)
                .error(placeholder)
                .dontAnimate()
                .diskCacheStrategy(DiskCacheStrategy.NONE);
        GlideApp.with(mContext).asBitmap().load(bitmap).apply(options).into(img);
    }

    /**
     * --------------------------加载图片File--------------------------------------
     */
    public static void loadingImg(Context mContext, File file, ImageView img,
                                  int placeholder) {
        GlideApp.with(mContext).clear(img);
        RequestOptions options = new RequestOptions().placeholder(placeholder)
                .error(placeholder)
                .dontAnimate()
                .diskCacheStrategy(DiskCacheStrategy.NONE);
        GlideApp.with(mContext).asFile().load(file).apply(options).into(img);
    }

    /**
     * --------------------------单独下载加载图片--------------------------------------
     */
    public static void loadingImg(final Context mContext, String url,
                                  final LoadingImageListener listener, int width, int height) {
        GlideApp.with(mContext)
                .asBitmap()
                .load(url)
                .into(new SimpleTarget<Bitmap>() {
                    @Override
                    public void onResourceReady(Bitmap resource, Transition<? super Bitmap> transition) {
                        listener.onResourceReady(resource);
                    }

                    @Override
                    public void onLoadFailed(@Nullable Drawable errorDrawable) {
                        super.onLoadFailed(errorDrawable);
//                        GlideApp.with(mContext).clear(this);
                        listener.onLoadFailed();
                    }
                });
    }

    /**
     * --------------------------加载网络普通图片-----------------------------------
     */
    public static void loadingImgSimple(Context mContext, String url, ImageView img) {
        GlideApp.with(mContext).load(url).into(img);
    }

    /**
     * @param mContext
     * @param url
     * @param img
     * @param placeholder
     * @param listener
     * @param withAnim
     * @param width
     * @param height
     */
    public static void loadingImgWithCommon(final Context mContext,
                                            String url, final ImageView img, final int placeholder,
                                            final LoadingImageListener listener, boolean withAnim, int width,
                                            int height, int imageloader_anim) {
        GlideApp.with(mContext).clear(img);
        RequestOptions options = new RequestOptions().placeholder(placeholder)
                .error(placeholder)
                .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC);
        RequestBuilder<Drawable> mRequestBuilder = GlideApp.with(mContext)
                .load(url);
        //glide v4动画有问题，先注销
//        if (withAnim) {
//            mRequestBuilder.transition(withCrossFade(200));
//        } else {
            options.dontAnimate();
//        }
        if (width != 0 && height != 0) {
            options.override(width, height);
        }
        mRequestBuilder.apply(options);
        if (listener != null) {
            mRequestBuilder.listener(new RequestListener<Drawable>() {
                @Override
                public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                    listener.onLoadFailed();
                    return false;
                }

                @Override
                public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                    listener.onResourceReady(resource);
                    return false;
                }
            });
        }
        mRequestBuilder.into(img);
    }

    /**
     * 非wifi环境下 不显示图片
     *
     * @param mContext
     * @param url
     * @param img
     * @param placeholder
     */
    public static void loadingImgOutofWifi(Context mContext, String url,
                                           final ImageView img, int placeholder, int imageloader_anim) {
        GlideApp.with(mContext).clear(img);
        RequestOptions options = new RequestOptions().placeholder(placeholder)
                .error(placeholder)
                .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC);
        GlideApp.with(mContext)
                .load(placeholder)
                .apply(options).into(img);
    }

    /**
     * listview 加载圆形图片
     *
     * @param mContext
     * @param url
     * @param imageView 直接ImageView
     * @param radius    半径，如果是圆形，半径即为宽高的一般，只需要圆角，自定义
     */
    public static void loadingCircleImage(final Context mContext, String url,
                                          final ImageView imageView, final int radius,
                                          final Drawable placeholder, final LoadingImageListener listener) {
        GlideApp.with(mContext).clear(imageView);
        imageView.setImageDrawable(placeholder);
        RequestOptions options = new RequestOptions().placeholder(placeholder)
                .error(placeholder).centerCrop();
        GlideApp.with(mContext).asBitmap().load(url).apply(options)
                .into(new BitmapImageViewTarget(imageView) {
                    @Override
                    protected void setResource(Bitmap resource) {
                        RoundedBitmapDrawable circularBitmapDrawable = RoundedBitmapDrawableFactory
                                .create(mContext.getResources(), resource);
                        circularBitmapDrawable.setCornerRadius(radius);
                        imageView.setImageDrawable(circularBitmapDrawable);
                        if (listener != null) {
                            listener.onResourceReady(circularBitmapDrawable);
                        }
                    }

                    @Override
                    public void onLoadFailed(@Nullable Drawable errorDrawable) {
                        super.onLoadFailed(errorDrawable);
                        imageView.setImageDrawable(placeholder);
                        if (listener != null) {
                            listener.onLoadFailed();
                        }
                    }
                });
    }

    /**
     * 清除图片缓存
     *
     * @param mContext
     */
    public static void clearMemory(final Context mContext) {
        GlideApp.get(mContext).clearMemory();
    }

    public static void clearDiskCache(final Context mContext) {
        GlideApp.get(mContext).clearDiskCache();
    }

}

File: LibraryImageLoader/src/main/java/com/android/library/CoreImageLoaderUtil.kt
package com.android.library

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.widget.ImageView
import com.android.library.imageloader.R
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.load.resource.gif.GifDrawable
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.BitmapImageViewTarget
import com.bumptech.glide.request.target.Target
import com.bumptech.glide.request.transition.Transition
import com.android.library.GlideApp

/**
 *@author chenjunwei
 *@desc Glide工具类
 *@date 2019/8/23
 */
@SuppressLint("CheckResult")
object CoreImageLoaderUtil {
    /* --------------------------默认图片-------------------------- */
    var loading_50 = R.drawable.default_logo_50

    /**
     * 加载图片监听接口
     */
    interface LoadingImageListener {
        /**
         * 图片加载成功
         * @param t
         * @param <T>
         */
        fun <T> onResourceReady(t: T)

        /**
         * 图片加载失败
         */
        fun onLoadFailed()
    }

    /*----------------------------加载图片---------------------------*/

    /**
     * 加载本地图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     */
    fun loadingImg(mContext: Context, resource: Any, img: ImageView) {
        loadingImg(mContext, resource, img, loading_50)
    }

    /**
     * 加载本地图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     * @param placeholder   默认加载图
     */
    fun loadingImg(mContext: Context, resource: Any, img: ImageView, placeholder: Any?) {
        loadingImg(mContext, resource, img, placeholder, placeholder, null, 0, 0)
    }

    /**
     * 加载本地图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     * @param placeholder   默认加载图
     * @param failResId   加载错误图
     * @param listener  加载监听
     * @param width     加载图片宽
     * @param height    加载图片高
     */
    fun loadingImg(
        mContext: Context, resource: Any, img: ImageView, placeholder: Any?,
        failResId: Any?, listener: LoadingImageListener?, width: Int, height: Int
    ) {
        loadingImg(mContext, resource, img, placeholder, failResId, listener, width, height, DiskCacheStrategy.ALL)
    }

    /**
     * 加载本地图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     * @param placeholder   默认加载图
     * @param failResId   加载错误图
     * @param listener  加载监听
     * @param width     加载图片宽
     * @param height    加载图片高
     */
    fun loadingImg(
        mContext: Context, resource: Any, img: ImageView, placeholder: Any?, failResId: Any?,
        listener: LoadingImageListener?, width: Int, height: Int, diskCacheStrategy: DiskCacheStrategy
    ) {
        GlideApp.with(mContext).clear(img)
        val options = RequestOptions()
        if (width != 0 && height != 0) {
            options.override(width, height)
        }
        setPlaceHolder(placeholder, failResId, options)

        GlideApp.with(mContext)
            .asBitmap()
            .load(resource)
            .apply(options)
            .diskCacheStrategy(diskCacheStrategy)
            .dontAnimate()
            .listener(object : RequestListener<Bitmap> {
                override fun onLoadFailed(e: GlideException?, model: Any?, target: Target<Bitmap>?, isFirstResource: Boolean): Boolean {
                    listener?.onLoadFailed()
                    return false
                }

                override fun onResourceReady(
                    resource: Bitmap?,
                    model: Any?,
                    target: Target<Bitmap>?,
                    dataSource: DataSource?,
                    isFirstResource: Boolean
                ): Boolean {
                    listener?.onResourceReady(resource)
                    return false
                }

            })
            .into(img)
    }

    /**
     * 加载本地图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param placeholder   默认加载图
     * @param failResId   加载错误图
     * @param listener  加载监听
     * @param width     加载图片宽
     * @param height    加载图片高
     */
    fun loadingImg(
        mContext: Context, resource: Any, placeholder: Any?, failResId: Any?,
        listener: LoadingImageListener?, width: Int, height: Int
    ) {
        val options = RequestOptions()
        if (width != 0 && height != 0) {
            options.override(width, height)
        }
        setPlaceHolder(placeholder, failResId, options)
        GlideApp.with(mContext)
            .asBitmap()
            .load(resource)
            .apply(options)
            .diskCacheStrategy(DiskCacheStrategy.ALL)
            .dontAnimate()
            .listener(object : RequestListener<Bitmap> {
                override fun onLoadFailed(e: GlideException?, model: Any?, target: Target<Bitmap>?, isFirstResource: Boolean): Boolean {
                    listener?.onLoadFailed()
                    return false
                }

                override fun onResourceReady(
                    resource: Bitmap?,
                    model: Any?,
                    target: Target<Bitmap>?,
                    dataSource: DataSource?,
                    isFirstResource: Boolean
                ): Boolean {
                    listener?.onResourceReady(resource)
                    return false
                }

            })
    }

    /**
     * 在非wifi环境下,不远程加载图片
     * 注：如果图片在内存或在磁盘缓存中，它会被展示出来。否则不展示；
     *
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     */
    fun loadingImgOutOfWifi(mContext: Context, resource: Any, img: ImageView) {
        loadingImgOutOfWifi(mContext, resource, img, 0, 0)
    }

    /**
     * 在非wifi环境下,不远程加载图片
     * 注：如果图片在内存或在磁盘缓存中，它会被展示出来。否则不展示；
     *
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     * @param placeholder   占位默认图片
     */
    fun loadingImgOutOfWifi(mContext: Context, resource: Any, img: ImageView, placeholder: Any?, failResId: Any?) {
        loadingImgOutOfWifi(mContext, resource, img, placeholder, failResId, 0, 0)
    }

    /**
     * 在非wifi环境下,不远程加载图片
     * 注：如果图片在内存或在磁盘缓存中，它会被展示出来。否则不展示；
     *
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     * @param placeholder   占位默认图片
     * @param width     加载图片宽
     * @param height    加载图片高
     */
    fun loadingImgOutOfWifi(
        mContext: Context, resource: Any, img: ImageView, placeholder: Any?,
        failResId: Any?, width: Int, height: Int
    ) {
        loadingImgOutOfWifi(mContext, resource, img, placeholder, failResId, width, height, 0)
    }

    /**
     * 在非wifi环境下,不远程加载图片
     * 注：如果图片在内存或在磁盘缓存中，它会被展示出来。否则不展示；
     *
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     * @param placeholder   占位默认图片
     * @param width     加载图片宽
     * @param height    加载图片高
     * @param imageLoaderAnim  是否需要动画
     */
    fun loadingImgOutOfWifi(
        mContext: Context, resource: Any, img: ImageView, placeholder: Any?,
        failResId: Any?, width: Int, height: Int, imageLoaderAnim: Int
    ) {
        GlideApp.with(mContext).clear(img)
        val options = RequestOptions()
        if (width != 0 && height != 0) {
            options.override(width, height)
        }
        options.diskCacheStrategy(DiskCacheStrategy.ALL)
            .onlyRetrieveFromCache(true)
        setPlaceHolder(placeholder, failResId, options)
        val mRequestBuilder = GlideApp.with(mContext).load(resource)
        if (imageLoaderAnim > 0) {
            mRequestBuilder.transition(DrawableTransitionOptions().transition(imageLoaderAnim))
        }
        mRequestBuilder.apply(options)
            .into(img)
    }

    /**
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     * @param imageLoaderAnim  是否需要动画
     */
    fun loadingImgWithAnim(mContext: Context, resource: Any, img: ImageView, imageLoaderAnim: Int) {
        loadingImgWithAnim(mContext, resource, img, 0, 0, imageLoaderAnim)
    }

    /**
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     * @param placeholder   占位默认图片
     * @param imageLoaderAnim  是否需要动画
     */
    fun loadingImgWithAnim(mContext: Context, resource: Any, img: ImageView, placeholder: Any?, failResId: Any?, imageLoaderAnim: Int) {
        loadingImgWithAnim(mContext, resource, img, placeholder, failResId, 0, 0, imageLoaderAnim)
    }

    /**
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     * @param placeholder   占位默认图片
     * @param width     加载图片宽
     * @param height    加载图片高
     * @param imageLoaderAnim  是否需要动画
     */
    fun loadingImgWithAnim(
        mContext: Context, resource: Any, img: ImageView, placeholder: Any?,
        failResId: Any?, width: Int, height: Int, imageLoaderAnim: Int
    ) {
        GlideApp.with(mContext).clear(img)
        val options = RequestOptions()
        if (width != 0 && height != 0) {
            options.override(width, height)
        }
        setPlaceHolder(placeholder, failResId, options)
        val mRequestBuilder = GlideApp.with(mContext).load(resource)
        if (imageLoaderAnim > 0) {
            mRequestBuilder.transition(DrawableTransitionOptions().transition(imageLoaderAnim))
        }
        mRequestBuilder.apply(options)
            .diskCacheStrategy(DiskCacheStrategy.ALL)
            .into(img)
    }

    /**
     * 加载圆形图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param imageView     目标view
     * @param radius    圆形图片半径
     */
    fun loadingCircleImage(mContext: Context, resource: Any, imageView: ImageView, radius: Int) {
        loadingCircleImage(mContext, resource, imageView, radius, loading_50)
    }

    /**
     * 加载圆形图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param imageView     目标view
     * @param radius    圆形图片半径
     */
    fun loadingCircleImage(mContext: Context, resource: Any, imageView: ImageView, radius: Int, placeholder: Any?) {
        loadingCircleImage(mContext, resource, imageView, radius, placeholder, 0, 0)
    }

    /**
     * 加载圆形图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param imageView     目标view
     * @param radius    圆形图片半径
     */
    fun loadingCircleImage(
        mContext: Context, resource: Any, imageView: ImageView, radius: Int,
        placeholder: Any?, width: Int, height: Int
    ) {
        loadingCircleImage(mContext, resource, imageView, radius, placeholder, placeholder, null, width, height, DiskCacheStrategy.ALL)
    }

    /**
     * 加载圆形图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param imageView     目标view
     * @param radius    圆形图片半径
     * @param placeholder   占位默认图片
     * @param listener  加载监听
     */
    fun loadingCircleImage(
        mContext: Context, resource: Any, imageView: ImageView, radius: Int, placeholder: Any?, failResId: Any?,
        listener: LoadingImageListener?, width: Int, height: Int, diskCacheStrategy: DiskCacheStrategy
    ) {
        GlideApp.with(mContext).clear(imageView)
        val options = RequestOptions()
        if (width != 0 && height != 0) {
            options.override(width, height)
        }
        setPlaceHolder(placeholder, failResId, options)
        GlideApp.with(mContext)
            .asBitmap()
            .load(resource)
            .apply(options)
            .diskCacheStrategy(diskCacheStrategy)
            .transforms(CenterCrop(), RoundedCorners(radius))
            .into(object : BitmapImageViewTarget(imageView) {
                override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
                    super.onResourceReady(resource, transition)
                    listener?.onResourceReady(resource)
                }

                override fun onLoadFailed(errorDrawable: Drawable?) {
                    super.onLoadFailed(errorDrawable)
                    listener?.onLoadFailed()
                }
            })
    }

    /*----------------------------加载gif图片---------------------------*/

    /**
     * 加载本地图片到指定view
     * @param mContext  上下文
     * @param resource   本地图片地址/网络地址/bitmap等等;具体看Glide load方法
     * @param img   目标view
     */
    fun loadingGifImg(mContext: Context, resource: Any, img: ImageView) {
        loadingGifImg(mContext, resource, img, 0, 0)
    }

    /**
     * 加载本地图片到指定view
     * @param mContext  上下文
     * @param resource   gif路径
     * @param img   目标view
     * @param placeholder   默认加载图
     * @param failResId   加载错误图
     */
    fun loadingGifImg(mContext: Context, resource: Any, img: ImageView, placeholder: Any?, failResId: Any?) {
        loadingGifImg(mContext, resource, img, placeholder, failResId, null)
    }

    fun loadingGifImg(
        mContext: Context, resource: Any, img: ImageView, placeholder: Any?,
        failResId: Any?, listener: LoadingImageListener?
    ) {
        loadingGifImg(mContext, resource, img, 0, 0, placeholder, failResId, listener)
    }

    /**
     * 加载本地图片到指定view
     * @param mContext  上下文
     * @param resource   gif路径
     * @param img   目标view
     * @param placeholder   默认加载图
     * @param failResId   加载错误图
     * @param listener  加载监听
     * @param width     加载图片宽
     * @param height    加载图片高
     */
    fun loadingGifImg(
        mContext: Context, resource: Any, img: ImageView, width: Int,
        height: Int, placeholder: Any?, failResId: Any?, listener: LoadingImageListener?
    ) {
        GlideApp.with(mContext).clear(img)
        val options = RequestOptions()
        if (width != 0 && height != 0) {
            options.override(width, height)
        }
        setPlaceHolder(placeholder, failResId, options)
        GlideApp.with(mContext)
            .asGif()
            .load(resource)
            .apply(options)
            .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
            .listener(object : RequestListener<GifDrawable> {
                override fun onLoadFailed(e: GlideException?, model: Any?, target: Target<GifDrawable>?, isFirstResource: Boolean): Boolean {
                    listener?.onLoadFailed()
                    return false
                }

                override fun onResourceReady(resource: GifDrawable?, model: Any?, target: Target<GifDrawable>?, dataSource: DataSource?, isFirstResource: Boolean): Boolean {
                    listener?.onResourceReady(resource)
                    return false
                }

            })
            .into(img)
    }


    /**
     * 设置默认、错误 加载图
     * @param placeholder   默认加载图
     * @param failResId     错误加载图
     * @param options
     */
    private fun setPlaceHolder(placeholder: Any?, failResId: Any?, options: RequestOptions) {
        if (null != placeholder) {
            if (placeholder is Drawable) {
                options.placeholder(placeholder)
            } else if (placeholder is Int) {
                if (placeholder > 0) {
                    options.placeholder(placeholder)
                }
            }
        }
        if (null != failResId) {
            if (failResId is Drawable) {
                options.placeholder(failResId)
            } else if (failResId is Int) {
                if (failResId > 0) {
                    options.placeholder(failResId)
                }
            }
        }
    }

    /**
     * 清除图片内存缓存
     * @param mContext
     */
    fun clearMemory(mContext: Context) {
        Glide.get(mContext).clearMemory()
    }

    /**
     * 清除图片磁盘缓存
     * @param mContext
     */
    fun clearDishCache(mContext: Context) {
        GlideApp.get(mContext).clearDiskCache()
    }

}